### PR TITLE
toolbox: add toolbox/gpu-operator/run_gpu_burn.sh script

### PR DESCRIPTION
**Pull request by kpouget**
* 4dd82ec - callback_plugins/human_log.py: correctly group 'task.get_path()' with its task logs

---

* 16dd890 - callback_plugins/human_log.py: improve readability

The task output now looks like this:

```
---

---
roles/nv_gpu/tasks/ci_checks.yml:4
TASK: nv_gpu : CRD check - Ci mode

<command> oc get crds/clusterpolicies.nvidia.com

<stdout> NAME                         CREATED AT
<stdout> clusterpolicies.nvidia.com   2021-03-08T09:27:30Z

Monday 08 March 2021  10:48:52 +0100 (0:00:01.961)       0:00:04.545 **********
```

or when it fails:
```
---

---
roles/nv_gpu/tasks/ci_checks.yml:15
TASK: nv_gpu : Ensure that NFD found nodes with GPU labels
----- FAILED ----
msg: non-zero return code

<command> (   oc get nodes -oname --ignore-not-found=false -l feature.node.kubernetes.io/pci-10de.present || oc get nodes -oname --ignore-not-found=false -l feature.node.kubernetes.io/pci-0302_10de.present || oc get nodes -oname --ignore-not-found=false -l feature.node.kubernetes.io/pci-0300_10de.present ) | grep .

----- FAILED ----

Monday 08 March 2021  10:51:38 +0100 (0:00:00.812)       0:00:05.870 **********
```

---

* 54c9a62 - callback_plugins/human_log.py: improve to handle ansible 'with_items' loops

---

* 8a0d4f5 - toolbox: add toolbox/gpu-operator/run_gpu_burn.sh script

---

* 8bfc7a0 - build/root/usr/local/bin/run: call gpu-burn in the CI tests

### CODE REVIEW

**Comment by kpouget** [comment]
Testing looks [good](
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift-psap_ci-artifacts/83/pull-ci-openshift-psap-ci-artifacts-master-gpu-operator-e2e/1368983259383336960/artifacts/gpu-operator-e2e/presubmit-master/build-log.txt) :)

---
roles/gpu_burn/tasks/main.yml:62
TASK: gpu_burn : Save the logs of the GPU burn Pods
msg: All items completedLOOP #0: item = ip-10-0-160-117<command> oc logs pod/gpu-burn-ip-10-0-160-117 -n default | grep -v "proc'd"<stdout> Running on 1 GPUs:
<stdout> GPU 0: Tesla T4 (UUID: GPU-d8b769e8-bc89-425a-8d07-c17b2d9e72b0)
<stdout> 
<stdout> GPU 0: Tesla T4 (UUID: GPU-d8b769e8-bc89-425a-8d07-c17b2d9e72b0)
<stdout> 	Summary at:   Mon Mar  8 19:03:07 UTC 2021
<stdout> 
<stdout> 	Summary at:   Mon Mar  8 19:03:14 UTC 2021
<stdout> 
<stdout> 	Summary at:   Mon Mar  8 19:03:21 UTC 2021
<stdout> 
<stdout> 	Summary at:   Mon Mar  8 19:03:28 UTC 2021
<stdout> 
<stdout> 	Summary at:   Mon Mar  8 19:03:35 UTC 2021
<stdout> 
<stdout> 	Summary at:   Mon Mar  8 19:03:42 UTC 2021
<stdout> 
<stdout> 	Summary at:   Mon Mar  8 19:03:49 UTC 2021
<stdout> 
<stdout> 	Summary at:   Mon Mar  8 19:03:55 UTC 2021Monday 08 March 2021  19:03:59 +0000 (0:00:00.571)       0:01:06.516 ********** 
---
```
I just need to find the way to remove the logging of every % of progress, but keep a bit more info than above
and I'll add a parameter to `toolbox/gpu-operator/run_gpu_burn.sh` to override the burn duration

**Comment by kpouget** [comment]
\hold

**Comment by kpouget** [comment]
The new feature:
- [x] has a documentation entry
  - `README.md`
- [x] defines a new playbook 
  - `playbooks/gpu-burn.yml`
- [x] defines a new role
  - `roles/gpu/burn`
- [x] has a `toolbox` entrypoint
  - ` toolbox/gpu-operator/run_gpu_burn.sh `
  - [x]  is included in the CI testing (`build/root/usr/local/bin/run `)

**Comment by kpouget** [comment]
hum, interesting, I got a test failure with the A100 in MIG mode:
```
<stdout> Running on 3 GPUs:
<stdout> GPU 0: A100-PCIE-40GB (UUID: GPU-d9322296-54da-ce5a-6330-3ca7707e0c5d)
<stdout>   MIG 3g.20gb Device 0: (UUID: MIG-GPU-d9322296-54da-ce5a-6330-3ca7707e0c5d/1/0)
<stdout>   MIG 3g.20gb Device 1: (UUID: MIG-GPU-d9322296-54da-ce5a-6330-3ca7707e0c5d/2/0)
<stdout> Failure during compute: Error in "SGEMM": CUBLAS_STATUS_EXECUTION_FAILED
<stdout> 37.0%  proc'd: -1 (0 Gflop/s)   errors: -1  (DIED!)  temps: 34 C
<stdout> 	Summary at:   Initialized device 0 with 20096 MB of memory (19869 MB available, using 17882 MB of it), using FLOATS
<stdout> Fri Mar 12 15:26:31 UTC 2021
<stdout> No clients are alive!  Aborting
<stdout> real	1m50.819s
<stdout> user	0m0.080s
<stdout> sys	0m0.029s
```

but it wasn't supposed to use 3 GPUs, only the 2 MIG devices ...

**Comment by kpouget** [comment]
@ArangoGutierrez the GPU burn playbook executes correctly, this PR is good for review / merging

```
---
roles/gpu_burn/tasks/main.yml:62
TASK: gpu_burn : Ensure that no GPU was faulty
msg: All items completed

LOOP #0: item = ip-10-0-174-79
msg: non-zero return code

<command> oc logs pod/gpu-burn-ip-10-0-174-79 -n default | grep FAULTY


Friday 12 March 2021  17:10:58 +0000 (0:00:00.909)       0:05:40.285 ********** 
---


---
roles/gpu_burn/tasks/main.yml:70
TASK: gpu_burn : Save the logs of the GPU burn Pods
msg: All items completed

LOOP #0: item = ip-10-0-174-79

<command> oc logs pod/gpu-burn-ip-10-0-174-79 -n default | grep -o "[^$(printf '\r')]*$"

<stdout> Running on 1 GPUs:
<stdout> GPU 0: Tesla T4 (UUID: GPU-c58253a6-e3f0-8e30-d38f-d9d310d4999f)
<stdout> GPU 0: Tesla T4 (UUID: GPU-c58253a6-e3f0-8e30-d38f-d9d310d4999f)
<stdout> 10.3%  proc'd: 7542 (4190 Gflop/s)   errors: 0   temps: 37 C 
<stdout> 	Summary at:   Fri Mar 12 17:06:01 UTC 2021
<stdout> 20.7%  proc'd: 15084 (4106 Gflop/s)   errors: 0   temps: 45 C 
<stdout> 	Summary at:   Fri Mar 12 17:06:32 UTC 2021
<stdout> 31.0%  proc'd: 22626 (4152 Gflop/s)   errors: 0   temps: 41 C 
<stdout> 	Summary at:   Fri Mar 12 17:07:03 UTC 2021
<stdout> 41.3%  proc'd: 30168 (4198 Gflop/s)   errors: 0   temps: 37 C 
<stdout> 	Summary at:   Fri Mar 12 17:07:34 UTC 2021
<stdout> 51.7%  proc'd: 36872 (4200 Gflop/s)   errors: 0   temps: 37 C 
<stdout> 	Summary at:   Fri Mar 12 17:08:05 UTC 2021
<stdout> 61.7%  proc'd: 44414 (4195 Gflop/s)   errors: 0   temps: 38 C 
<stdout> 	Summary at:   Fri Mar 12 17:08:35 UTC 2021
<stdout> 72.3%  proc'd: 52794 (4173 Gflop/s)   errors: 0   temps: 39 C 
<stdout> 	Summary at:   Fri Mar 12 17:09:07 UTC 2021
<stdout> 82.7%  proc'd: 60336 (4143 Gflop/s)   errors: 0   temps: 41 C 
<stdout> 	Summary at:   Fri Mar 12 17:09:38 UTC 2021
<stdout> 93.3%  proc'd: 67878 (4113 Gflop/s)   errors: 0   temps: 44 C 
<stdout> 	Summary at:   Fri Mar 12 17:10:10 UTC 2021
<stdout> 100.0%  proc'd: 72906 (4115 Gflop/s)   errors: 0   temps: 47 C 
<stdout> Killing processes.. done
<stdout> Tested 1 GPUs:
<stdout> 	GPU 0: OK
<stdout> real	5m1.174s
<stdout> user	3m1.974s
<stdout> sys	1m57.745s

Friday 12 March 2021  17:10:59 +0000 (0:00:00.939)       0:05:41.225 ********** 
```

and full GPU burn logs are there:

https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift-psap_ci-artifacts/83/pull-ci-openshift-psap-ci-artifacts-master-gpu-operator-e2e/1370401941166755840/artifacts/gpu-operator-e2e/presubmit-master/artifacts/gpu-commit-ci/extra_logs/gpu_burn.ip-10-0-174-79.logs

File: callback_plugins/human_log.py [hunk 1]
```diff
@@ -18,37 +19,55 @@
...
         if ignore_errors is not None:
             self._display.display(f"==> FAILED | ignore_errors={ignore_errors}", color=color)
 
 
         try: self._display.display(f"msg: {result._result['msg']}", color=color)
         except KeyError: pass
 
-        try:
-            cmd = result._result['cmd']
+        if "results" in result._result:
```
Comment (ArangoGutierrez) [inline callback_plugins/human_log.py:30]: for things like `result._result` is that I don't get python ;P

File: roles/gpu_burn/files/gpu_burn_pod.yml [hunk 2]
```diff
@@ -0,0 +1,31 @@
+apiVersion: v1
+kind: Pod
```
Comment (ArangoGutierrez) [inline roles/gpu_burn/files/gpu_burn_pod.yml:2]: do you think we could use `job` instead of `pod`?

**Review by ArangoGutierrez — APPROVED** [review]
Loks good to me with some comments but I am not blocking the PR

/lgtm
/hold

File: roles/gpu_burn/files/gpu_burn_pod.yml (hunk 2 above)
Comment (kpouget) [reply roles/gpu_burn/files/gpu_burn_pod.yml:2]: ack, we keep the `Pod` for now, but one day we'll turn them into `Jobs`!

File: callback_plugins/human_log.py (hunk 1 above)
Comment (kpouget) [reply callback_plugins/human_log.py:30]: `_result` means that this field is internal to the class of `result`, it's only a naming convention (`__result` is stronger, and the languages helps hiding it by mangling it into something like `_classname_result`)

**Pull request merged**